lshw.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "LSHW" "1" "09 May 2004" "$Id$" ""
.SH NAME
lshw \- list hardware
.SH SYNOPSIS

\fBlshw\fR [ \fB-version\fR ]


\fBlshw\fR [ \fB-help\fR ]


\fBlshw\fR [ \fB-html\fR | \fB-short\fR | \fB-xml\fR ] [ \fB-class \fIclass\fB\fR\fI ...\fR ] [ \fB-disable \fItest\fB\fR\fI ...\fR ] [ \fB-enable \fItest\fB\fR\fI ...\fR ]

.SH "DESCRIPTION"
.PP

\fBlshw\fR
is a small tool to extract detailed information on the hardware
configuration of the machine. It can report exact memory
configuration, firmware version, mainboard configuration, 
CPU version
and speed, cache configuration, bus speed, etc. on 
DMI-capable x86 or IA-64
systems and on some PowerPC
machines (PowerMac G4 is known to work).
.PP
It currently supports DMI (x86 and IA-64 only), OpenFirmware device tree (PowerPC only),
PCI/AGP, CPUID (x86), IDE/ATA/ATAPI, PCMCIA (only tested on x86) and SCSI.
.PP
.TP
\fB-version\fR
Displays the version of \fBlshw\fR and exits.
.TP
\fB-help\fR
Displays the available command line options and quits.
.TP
\fB-html\fR
Outputs the device tree as an HTML page.
.TP
\fB-xml\fR
Outputs the device tree as an XML tree.
.TP
\fB-short\fR
Outputs the device tree showing hardware paths, very much like the output of HP-UX's \fBioscan\fR.
.TP
\fB-class \fIclass\fB\fR
Only show the given class of hardware. \fIclass\fR can be found using \fBlshw -short\fR.
.TP
\fB-C \fIclass\fB\fR
Alias for \fB-class\fR \fIclass\fR.
.TP
\fB-enable \fItest\fB\fR
.TP
\fB-disable \fItest\fB\fR
Enables or disables a test. \fItest\fR can be \fBdmi\fR (for DMI/SMBIOS extensions), \fBdevice-tree\fR (for OpenFirmware device tree), \fBspd\fR (for memory Serial Presence Detect), \fBmemory\fR (for memory-size guessing heuristics), \fBcpuinfo\fR (for kernel-reported CPU detection), \fBcpuid\fR (for CPU detection), \fBpci\fR (for PCI/AGP access), \fBisapnp\fR (for ISA PnP extensions), \fBpcmcia\fR (for PCMCIA/PCCARD), \fBide\fR (for IDE/ATAPI), \fBscsi\fR (for SCSI) or \fBnetwork\fR (for network interfaces detection).
.SH "BUGS"
.PP
\fBlshw\fR currently does not detect 
USB or Firewire(IEEE1394) devices.
.PP
Not all architectures supported by GNU/Linux are fully supported (e.g.
CPU detection).
.PP
"Virtual" SCSI interfaces used for USB mass storage and SCSI emulation over IDE are not reported correctly yet.
.SH "NOTES"
.PP
\fBlshw\fR must be run as super user or it will only report
partial information.
.SH "FILES"
.PP
.TP
\fB/usr/local/share/pci.ids\fR
.TP
\fB/usr/share/pci.ids\fR
.TP
\fB/etc/pci.ids\fR
.TP
\fB/usr/share/hwdata/pci.ids\fR
A list of all known PCI ID's (vendors,  devices, classes and subclasses).
.TP
\fB/proc/bus/pci/*\fR
Used to access the configuration of installed PCI busses and devices.
.TP
\fB/proc/ide/*\fR
Used to access the configuration of installed IDE busses and devices.
.TP
\fB/proc/scsi/*, /dev/sg*\fR
Used to access the configuration of installed SCSI devices.
.TP
\fB/dev/cpu/*/cpuid\fR
Used on x86 platforms to access CPU-specific configuration.
.TP
\fB/proc/device-tree/*\fR
Used on PowerPC platforms to access OpenFirmware configuration.
.TP
\fB/sys/*\fR
Used on 2.6 kernels to access hardware/driver configuration information.
.SH "EXAMPLES"
.PP
.TP
\fBlshw -short\fR
Lists hardware in a compact format.
.TP
\fBlshw -class disk -class storage\fR
Lists all disks and storage controllers in the system.
.TP
\fBlshw -html -class network\fR
Lists all network interfaces in HTML.
.TP
\fBlshw -disable dmi\fR
Don't use DMI to detect hardware.
.SH "SEE ALSO"
.PP
\fI/proc/*\fR, \fBlinuxinfo\fR(1), \fBlspci\fR(8), \fBlsusb\fR(8)
.SH "COPYING"
.PP
\fBlshw\fR is distributed under the GNU GENERAL PUBLIC LICENSE (GPL) version 2.
.SH "AUTHOR"
.PP
\fBlshw\fR is maintained by Lyonel Vincent
<lyonel@ezix.org>.
.SH "OTHER INFO"
.PP
The webpage for \fBlshw\fR is at 
 <URL:http://ezix.org/software/lshw.html>


docs/Changelog
*  B.02.02
  added reporting of battery-related information using DMI
  added reporting of hardware security (DMI): admin/user/power-on passwords
  FIX: worked around the disparition of /proc/ide/ide*/config on 2.6.10
  new "businfo" output style
  first try to define an XML schema (cf. XSD file in docs/)
  renamed lm capability to x86-64 (easier to understand)
  portability fixes for RedHat 7.3
  updated USB IDs
  documentation updates
  added support for /proc/cpuinfo on x86-64
  code cleanup
*  B.02.01
  better support for emulated SCSI devices: USB mass storage devices are now properly handled
  added a "go-up" button to the GUI
  lshw now reports x86-64 capability
  minor fixes to give better descriptions of capabilities
  automatically create ghost (removed after use) /dev/sgxx entries when needed (affects Fedora Core 3 and other distributions that don't include these entries)
  USB code clean-up
  portability fixes for Gentoo Linux
*  B.02.00
  various fixes
  added a GTK+ GUI
* A.01.09
  code cleanup
  added support for USB devices
  various fixes
  updated list of recognised PCI devices
  added list of recognised USB devices
  source code reorganisation
  preparation for the development of a GTK+-based GUI
* A.01.08
  code cleanup
  added support for static and compressed (UPX) static builds
  added support for HT (hyper-threading); lshw now reports "virtual" CPUs
  DMI fixes (thanks to Erwan Velu from MandrakeSoft)
  XML output filtering
  packaging system fixes (thanks to Oron Peled)
  updated list of recognised PCI devices
* A.01.07
  fixed annoying IDE-related bug that caused un-necessary kernel error messages
  portability fixes
  migrated lshw's source code to Subversion
  build system updates
  more work on sysfs
  report WLAN capabilities
  report speed of network links
  bugfixes
* A.01.06
  code cleanup
  support for EFI-based (IA-64) systems, thanks to Bruno Cornec from HP
  the output of lshw can now be restricted to certain classes of hardware (disks, memory, etc.)
  preliminary support for 2.6 kernel using sysfs
  very preliminary support for USB devices (disabled by default)
  fixed typos
* A.01.05
  code cleanup
  PCI and PCMCIA now correctly report resources (ports, memory and IRQs)
  support for framebuffer devices
  report detailed DIMM info on PowerPC machines
  fixed compilation problem for older distributions
* A.01.04
  beginning of resources reporting
  bug fixes for multi-host bridge machines (servers, PowerMacs)
  fixes for OpenFirmware machines (PowerMacs)
  SCSI-over-IDE fixes (now connected to the right IDE device)
  new ability to disable some tests
  ISA PnP is now disabled by default (was too long)
* A.01.03
  new ioscan (HP-UX) -like output format
  added detection of ISA Plug & Play cards on x86 machines
  lshw now reports more detailed information on network interfaces
* A.01.02
  cosmetic changes
  added detection of Apple model names for Macs
  1st try of SPD memory DIMM support
  lshw now reports information on network interfaces
* A.01.01
  fixed memory reporting for systems with more than 2GB of memory
* A.01.00
  minor bug fixes
  support for more platforms (PA-RISC, IA-64 and Alpha)
  lshw can now generate XML
* T.00.07
  minor bug fixes
  lshw now generates correct HTML (thanks to Ola Lundqvist)
  SCSI support
  added a manpage (thanks to Helge Kreutzmann)
* T.00.06
  minor bug fixes
  PCMCIA support
  support for device claiming
  preliminary support for non-PCI IDE controllers/devices
* T.00.05
  code clean-up
  probing of IDE devices
* T.00.04
  activated HTML mode
  reporting of video memory
  bug fixes
* T.00.03
  fixes for GCC 3.2
* T.00.02
  second public test release
  PCI busses listing
* T.00.01
  first public test release


src/lshw.cc
#include "hw.h"
#include "print.h"
#include "main.h"
#include "version.h"
#include "options.h"
#include "osutils.h"
#include "config.h"

#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <iostream>

__ID("@(#) $Id$");

void usage(const char *progname)
{
  fprintf(stderr, "Hardware Lister (lshw) - %s\n", getpackageversion());
  fprintf(stderr, "usage: %s [-format] [-options ...]\n", progname);
  fprintf(stderr, "       %s -version\n", progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "\t-version        print program version (%s)\n", getpackageversion());
  fprintf(stderr, "\nformat can be\n");
  fprintf(stderr, "\t-html           output hardware tree as HTML\n");
  fprintf(stderr, "\t-xml            output hardware tree as XML\n");
  fprintf(stderr, "\t-short          output hardware paths\n");
  fprintf(stderr, "\t-businfo        output bus information\n");
  if(getenv("DISPLAY") && exists(SBINDIR"/gtk-lshw"))
    fprintf(stderr, "\t-X              use graphical interface\n");
  fprintf(stderr, "\noptions can be\n");
  fprintf(stderr,
    "\t-class CLASS    only show a certain class of hardware\n");
  fprintf(stderr, "\t-C CLASS        same as '-class CLASS'\n");
  fprintf(stderr, "\t-c CLASS        same as '-class CLASS'\n");
  fprintf(stderr,
    "\t-disable TEST   disable a test (like pci, isapnp, cpuid, etc. )\n");
  fprintf(stderr,
    "\t-enable TEST    enable a test (like pci, isapnp, cpuid, etc. )\n");
  fprintf(stderr, "\t-quiet          don't display status\n");
  fprintf(stderr, "\t-sanitize       sanitize output (remove sensitive information like serial numbers, etc.)\n");
  fprintf(stderr, "\n");
}


void status(const char *message)
{
  static size_t lastlen = 0;

  if(enabled("output:quiet") || disabled("output:verbose"))
    return;

  if (isatty(2))
  {
    fprintf(stderr, "\r");
    for (size_t i = 0; i < lastlen; i++)
      fprintf(stderr, " ");
    fprintf(stderr, "\r%s\r", message);
    fflush(stderr);
    lastlen = strlen(message);
  }
}


int main(int argc,
char **argv)
{
  disable("isapnp");

  disable("output:xml");
  disable("output:html");
  disable("output:hwpath");
  disable("output:businfo");
  disable("output:X");
  disable("output:quiet");
  disable("output:sanitize");

// define some aliases for nodes classes
  alias("disc", "disk");
  alias("cpu", "processor");
  alias("lan", "network");
  alias("net", "network");
  alias("video", "display");
  alias("sound", "multimedia");
  alias("modem", "communication");

  if (!parse_options(argc, argv))
  {
    usage(argv[0]);
    exit(1);
  }

  while (argc >= 2)
  {
    bool validoption = false;

    if (strcmp(argv[1], "-version") == 0)
    {
      const char *latest = checkupdates();
      printf("%s\n", getpackageversion());
      if(latest)
      {
        if(strcmp(latest, getpackageversion()) != 0)
          fprintf(stderr, "the latest version is %s\n", latest);
      }
      exit(0);
    }

    if (strcmp(argv[1], "-help") == 0 || strcmp(argv[1], "--help") == 0)
    {
      usage(argv[0]);
      exit(0);
    }

    if (strcmp(argv[1], "-verbose") == 0)
    {
      disable("output:quiet");
      enable("output:verbose");
      validoption = true;
    }

    if (strcmp(argv[1], "-quiet") == 0)
    {
      disable("output:verbose");
      enable("output:quiet");
      validoption = true;
    }

    if (strcmp(argv[1], "-xml") == 0)
    {
      enable("output:xml");
      validoption = true;
    }

    if (strcmp(argv[1], "-html") == 0)
    {
      enable("output:html");
      validoption = true;
    }

    if (strcmp(argv[1], "-short") == 0)
    {
      enable("output:hwpath");
      validoption = true;
    }

    if (strcmp(argv[1], "-businfo") == 0)
    {
      enable("output:businfo");
      validoption = true;
    }

    if (strcmp(argv[1], "-X") == 0)
    {
      enable("output:X");
      validoption = true;
    }

    if ((strcmp(argv[1], "-sanitize") == 0) ||
       (strcmp(argv[1], "-sanitise") == 0))
    {
      enable("output:sanitize");
      validoption = true;
    }

    if(validoption)
    {	/* shift */
      memmove(argv+1, argv+2, (argc-1)*(sizeof(argv[0])));
      argc--;
    }
    else
    {
      usage(argv[0]);
      exit(1);
    }
  }

  if (argc >= 2)
  {
    usage(argv[0]);
    exit(1);
  }

  if(enabled("output:X")) execl(SBINDIR"/gtk-lshw", SBINDIR"/gtk-lshw", NULL);

  if (geteuid() != 0)
  {
    fprintf(stderr, "WARNING: you should run this program as super-user.\n");
  }

  {
    hwNode computer("computer",
      hw::system);

    scan_system(computer);

    if (enabled("output:hwpath"))
      printhwpath(computer);
    else
    if (enabled("output:businfo"))
      printbusinfo(computer);
    else
    {
      if (enabled("output:xml"))
        cout << computer.asXML();
      else
        print(computer, enabled("output:html"));
    }
  }

  return 0;
}
